File: abc225/c/main.py
def main() -> bool:
    N, M = [int(x) for x in input().split(' ')]
    bss = [[int(x) - 1 for x in input().split(' ')] for _ in range(N)]

    start_y = bss[0][0] // 7
    start_x = bss[0][0] % 7

    for diff_y, bs in enumerate(bss):
        for diff_x, b in enumerate(bs):
            expected_y = b // 7
            expected_x = b % 7

            # print(start_y, start_x, b, diff_y, diff_x, expected_y, expected_x)

            if (start_y + diff_y != expected_y) or \
                    (start_x + diff_x != expected_x):
                return False

    return True


if __name__ == '__main__':
    if main():
        print('Yes')
    else:
        print('No')

File: abc225/c/main.rs
pub fn solve(n: usize, m: usize, bss: Vec<Vec<i32>>) -> bool {
    let mut yss: Vec<Vec<i32>> = vec![vec![-1; m]; n];
    let mut xss: Vec<Vec<i32>> = vec![vec![-1; m]; n];

    for (i, bs) in bss.iter().enumerate() {
        for (j, b) in bs.iter().enumerate() {
            yss[i][j] = (*b + 6) / 7;
            xss[i][j] = ((*b - 1) % 7) + 1;
        }
    }

    for i in 0..n {
        for j in 0..m {
            if i > 0 {
                if yss[i][j] != yss[i - 1][j] + 1 {
                    return false
                }
                if xss[i][j] != xss[i - 1][j] {
                    return false
                }
            }
            if j > 0 {
                if yss[i][j] != yss[i][j - 1] {
                    return false
                }
                if xss[i][j] != xss[i][j - 1] + 1 {
                    return false
                }
            }
        }
    }

    return true
}

pub fn main() {
    let (n, m): (usize, usize) = {
        let mut s = String::new();
        std::io::stdin().read_line(&mut s).unwrap();
        let xs: Vec<_> = s.trim().split(' ').map(|x| x.parse().unwrap()).collect();
        (xs[0], xs[1])
    };

    let bss: Vec<Vec<i32>> = {
        let mut acc = vec![];
        for _ in 0..n {
            let mut s = String::new();
            std::io::stdin().read_line(&mut s).unwrap();
            let xs: Vec<i32> = s.trim().split(' ').map(|x| x.parse().unwrap()).collect();
            acc.push(xs)
        }
        acc
    };

    if solve(n, m, bss) {
        println!("Yes");
    } else {
        println!("No");
    }
}
